woocommerce/woocommerce
# Product Gallery Before/After Init Events and adjustments to ProductGallery object to use custom arguments for Flexslider

### All Submissions:






### Changes proposed in this Pull Request:

Adds Events when looping through all occurrences of `.woocommerce-product-gallery` before and after running `$.fn.wc_product_gallery()`.

The `ProductGallery` object's `initFlexslider` method now also pulls in more values from `args` parameter, which means that if a 3rd party integration wanted to run `$.fn.wc_product_gallery()` with some custom options it can be done without overwriting the globally scoped `wc_single_product_params` object and potentially affecting the original Product Gallery.

### How to test the changes in this Pull Request:

1. Try viewing a Single Product in a Theme with `add_theme_support( 'wc-product-gallery-slider' );` added and check if the Product Gallery still loads correctly.
2. Try using the `woocommerce_single_product_carousel_options` Filter to ensure that the Product Gallery still works correctly when new options for the default Product Gallery are provided via PHP.
3. Try using the new `wc-product-gallery-before-init` and `wc-product-gallery-after-init` Events to run JavaScript before or after each `.woocommerce-product-gallery` becomes a Flexslider.
4. Try running the `$.fn.wc_product_gallery()` on a Product Gallery with a different CSS Selector using a custom `wc_single_product_params` object passed in (This will need to be ran at either one of the two new events, or at `$( document ).ready()` in order to ensure that the `$.fn.wc_product_gallery()` method is available)

### Other information:


I wanted to add a second Flexslider below the Product Gallery that also controls it, similar to http://flexslider.woothemes.com/thumbnail-slider.html

However, I could not reliably hook in before the Product Gallery way created in order to create the Flexslider for the bottom carousel before the default Product Gallery was. This was mainly due to how `$.fn.wc_product_gallery()` was loaded, so I had to wait until `$( document ).ready()` and by then, it was too late to get ahead of the default Product Gallery in JS. 

By using this newly included `wc-product-gallery-before-init` Event, I both have access to the `$.fn.wc_product_gallery()` method and I can get in there just before the default Product Gallery loads. With these changes, I can also provide a custom `wc_single_product_params` object to `$.fn.wc_product_gallery()` and have it use that when setting up Flexslider, PhotoSwipe, and Zoom for the new Product Gallery without overwriting the global `wc_single_product_params`, which could lead to some unintended consequences. 

It is difficult to tell in my low-framerate screen capture without a cursor showing, but first I let the bottom Flexslider automatically progress, then I click the images within it which changes the active slide in the default Product Gallery, then I click the image in the default Product Gallery which brings up PhotoSwipe. In order to get the slideshow to progress automatically on the bottom one, it had to be initialized first which was not possible without these changes.

![image](https://i.imgur.com/sYBSiaU.gif)

  - N/A, I don't believe. Seems all tests are for PHP. 
  - N/A, I don't believe. But I have tested this code against the functionality I wanted to use with this and it does work.



### Changelog entry

> Enter a short summary of all changes on this Pull Request. This will appear in the changelog if accepted.

* Adds `wc-product-gallery-before-init` and `wc-product-gallery-after-init` Events to run JavaScript before or after each `.woocommerce-product-gallery` becomes a Flexslider.
* Allows `$.fn.wc_product_gallery()` to be used by 3rd party integrations to initialize a Flexslider in a similar fashion to the Product Gallery and will now use a custom `wc_single_product_params` to be passed into it for changing options without potentially affecting the original Product Gallery.
